// hexo-util/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.wordWrap = exports.url_for = exports.unescapeHTML = exports.truncate = exports.tocObj = exports.stripIndent = exports.stripHTML = exports.spawn = exports.slugize = exports.relative_url = exports.prismHighlight = exports.prettyUrls = exports.Permalink = exports.Pattern = exports.isExternalLink = exports.htmlTag = exports.highlight = exports.createSha1Hash = exports.hash = exports.gravatar = exports.full_url_for = exports.escapeRegExp = exports.escapeHTML = exports.escapeDiacritic = exports.encodeURL = exports.deepMerge = exports.decodeURL = exports.Color = exports.camelCaseKeys = exports.CacheStream = exports.Cache = void 0;
var cache_1 = require("./cache");
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return __importDefault(cache_1).default; } });
var cache_stream_1 = require("./cache_stream");
Object.defineProperty(exports, "CacheStream", { enumerable: true, get: function () { return __importDefault(cache_stream_1).default; } });
var camel_case_keys_1 = require("./camel_case_keys");
Object.defineProperty(exports, "camelCaseKeys", { enumerable: true, get: function () { return __importDefault(camel_case_keys_1).default; } });
var color_1 = require("./color");
Object.defineProperty(exports, "Color", { enumerable: true, get: function () { return __importDefault(color_1).default; } });
var decode_url_1 = require("./decode_url");
Object.defineProperty(exports, "decodeURL", { enumerable: true, get: function () { return __importDefault(decode_url_1).default; } });
var deep_merge_1 = require("./deep_merge");
Object.defineProperty(exports, "deepMerge", { enumerable: true, get: function () { return __importDefault(deep_merge_1).default; } });
var encode_url_1 = require("./encode_url");
Object.defineProperty(exports, "encodeURL", { enumerable: true, get: function () { return __importDefault(encode_url_1).default; } });
var escape_diacritic_1 = require("./escape_diacritic");
Object.defineProperty(exports, "escapeDiacritic", { enumerable: true, get: function () { return __importDefault(escape_diacritic_1).default; } });
var escape_html_1 = require("./escape_html");
Object.defineProperty(exports, "escapeHTML", { enumerable: true, get: function () { return __importDefault(escape_html_1).default; } });
var escape_regexp_1 = require("./escape_regexp");
Object.defineProperty(exports, "escapeRegExp", { enumerable: true, get: function () { return __importDefault(escape_regexp_1).default; } });
var full_url_for_1 = require("./full_url_for");
Object.defineProperty(exports, "full_url_for", { enumerable: true, get: function () { return __importDefault(full_url_for_1).default; } });
var gravatar_1 = require("./gravatar");
Object.defineProperty(exports, "gravatar", { enumerable: true, get: function () { return __importDefault(gravatar_1).default; } });
var hash_1 = require("./hash");
Object.defineProperty(exports, "hash", { enumerable: true, get: function () { return hash_1.hash; } });
Object.defineProperty(exports, "createSha1Hash", { enumerable: true, get: function () { return hash_1.createSha1Hash; } });
var highlight_1 = require("./highlight");
Object.defineProperty(exports, "highlight", { enumerable: true, get: function () { return __importDefault(highlight_1).default; } });
var html_tag_1 = require("./html_tag");
Object.defineProperty(exports, "htmlTag", { enumerable: true, get: function () { return __importDefault(html_tag_1).default; } });
var is_external_link_1 = require("./is_external_link");
Object.defineProperty(exports, "isExternalLink", { enumerable: true, get: function () { return __importDefault(is_external_link_1).default; } });
var pattern_1 = require("./pattern");
Object.defineProperty(exports, "Pattern", { enumerable: true, get: function () { return __importDefault(pattern_1).default; } });
var permalink_1 = require("./permalink");
Object.defineProperty(exports, "Permalink", { enumerable: true, get: function () { return __importDefault(permalink_1).default; } });
var pretty_urls_1 = require("./pretty_urls");
Object.defineProperty(exports, "prettyUrls", { enumerable: true, get: function () { return __importDefault(pretty_urls_1).default; } });
var prism_1 = require("./prism");
Object.defineProperty(exports, "prismHighlight", { enumerable: true, get: function () { return __importDefault(prism_1).default; } });
var relative_url_1 = require("./relative_url");
Object.defineProperty(exports, "relative_url", { enumerable: true, get: function () { return __importDefault(relative_url_1).default; } });
var slugize_1 = require("./slugize");
Object.defineProperty(exports, "slugize", { enumerable: true, get: function () { return __importDefault(slugize_1).default; } });
var spawn_1 = require("./spawn");
Object.defineProperty(exports, "spawn", { enumerable: true, get: function () { return __importDefault(spawn_1).default; } });
var strip_html_1 = require("./strip_html");
Object.defineProperty(exports, "stripHTML", { enumerable: true, get: function () { return __importDefault(strip_html_1).default; } });
var strip_indent_1 = require("./strip_indent");
Object.defineProperty(exports, "stripIndent", { enumerable: true, get: function () { return __importDefault(strip_indent_1).default; } });
var toc_obj_1 = require("./toc_obj");
Object.defineProperty(exports, "tocObj", { enumerable: true, get: function () { return __importDefault(toc_obj_1).default; } });
var truncate_1 = require("./truncate");
Object.defineProperty(exports, "truncate", { enumerable: true, get: function () { return __importDefault(truncate_1).default; } });
var unescape_html_1 = require("./unescape_html");
Object.defineProperty(exports, "unescapeHTML", { enumerable: true, get: function () { return __importDefault(unescape_html_1).default; } });
var url_for_1 = require("./url_for");
Object.defineProperty(exports, "url_for", { enumerable: true, get: function () { return __importDefault(url_for_1).default; } });
var word_wrap_1 = require("./word_wrap");
Object.defineProperty(exports, "wordWrap", { enumerable: true, get: function () { return __importDefault(word_wrap_1).default; } });
